aten/src/THC/THCReduce.cuh

```diff
@@ -158,8 +158,8 @@ __global__ void kernelReduceNoncontigDim_shared
   IndexType sliceIndex  = blockIdx.x*blockDim.x + threadIdx.x;
 
   __shared__ int isLastBlockDone;
-  __shared__ AccT local_reduce[THC_NONCONTIG_REDUCE_BLOCK_SIZE];
-  AccT* shmem = &local_reduce[threadIdx.x + threadIdx.y*blockDim.x];
+  extern __shared__ char local_reduce[];
+  AccT* shmem = &(reinterpret_cast<AccT*>(local_reduce))[threadIdx.x + threadIdx.y*blockDim.x];
```

Comment:
are there any issues if we dont check at compile-time the size of the shared memory statically? if older archs dont have `THC_NONCONTIG_REDUCE_BLOCK_SIZE` sized shared memory per block, compilation will fail right?

Reply:
I switched to dynamic shared memory earlier to avoid some other warning about initialization with static shared memory.
Now changing the variable to `char` instead of `AccT`, I don't really need dynamic shared memory.

For some strange reason, I'm having issue with misaligned address after switching back to static shared memory. Let me dig it a little bit more here.

Reply:
Both for static and dynamic shared memory you should be taking alignment into account with `__align__` https://stackoverflow.com/questions/27570552/templated-cuda-kernel-with-dynamic-shared-memory

Reply:
@soumith no, unfortunately it will happily compile, and it will be a runtime error.

Reply:
Cleaning the constructor and switching back to static shared memory does the trick. I'll switch back to static so we can have a compile time other than a runtime error.
Will push once the build/test finishes.